rp2040-hal/src/i2c/controller/eh1_0_async.rs

```diff
@@ -0,0 +1,364 @@
+use eh_async_1_0_alpha::i2c::{AddressMode, Operation};
+use fugit::HertzU32;
+use rp2040_pac::RESETS;
+
+use crate::{
+    i2c::{Async, Controller, Error, ValidPinScl, ValidPinSda, I2C},
+    pac::i2c0::RegisterBlock,
+    resets::SubsystemReset,
+};
+
+struct PollI2c<'i2c, A, B, F, G, H, T>
+where
+    F: FnMut(&mut I2C<A, B, Controller<Async>>) -> Poll<T>,
+    G: FnMut(&mut I2C<A, B, Controller<Async>>),
+    H: FnMut(&mut I2C<A, B, Controller<Async>>),
+{
+    i2c: &'i2c mut I2C<A, B, Controller<Async>>,
+    is_ready: F,
+    setup_flags: G,
+    cancel: H,
+    running: bool,
+    // captures F's return type.
+    phantom: PhantomData<T>,
+}
+
+impl<'i2c, A, B, F, G, H, T> future::Future for PollI2c<'i2c, A, B, F, G, H, T>
+where
+    F: FnMut(&mut I2C<A, B, Controller<Async>>) -> Poll<T>,
+    G: FnMut(&mut I2C<A, B, Controller<Async>>),
+    H: FnMut(&mut I2C<A, B, Controller<Async>>),
+{
+    type Output = T;
+
+    fn poll(self: core::pin::Pin<&mut Self>, cx: &mut core::task::Context<'_>) -> Poll<T> {
+        // SAFETY: We are not moving anything.
+        let PollI2c {
+            ref mut i2c,
+            ref mut is_ready,
+            ref mut setup_flags,
+            ref mut running,
+            ..
+        } = unsafe { self.get_unchecked_mut() };
+        let r = (is_ready)(i2c);
+        if r.is_pending() {
+            if let Some(on_pending) = i2c.mode.m.on_pending {
+                on_pending(cx.waker().clone());
+                (setup_flags)(i2c);
+            } else {
+                // always ready to scan
+                cx.waker().wake_by_ref();
+            }
+        } else {
+            *running = false;
+        }
+        r
+    }
+}
+impl<'i2c, A, B, F, G, H, T> Drop for PollI2c<'i2c, A, B, F, G, H, T>
+where
+    F: FnMut(&mut I2C<A, B, Controller<Async>>) -> Poll<T>,
+    G: FnMut(&mut I2C<A, B, Controller<Async>>),
+    H: FnMut(&mut I2C<A, B, Controller<Async>>),
+{
+    fn drop(&mut self) {
+        if self.running {
+            (self.cancel)(self.i2c);
+        }
+    }
+}
+
+/// This macro blocks polling on the `$is_ready` function. When it is not ready,
+/// The a waker is passed to the `on_pending` callback and `$setup_flags` is called to enable
+/// the expected interrupt flag.
+macro_rules! block_on {
+    ($me:ident, $is_ready:ident ($($arg:expr),*), $setup_flags:ident ($($arg2:expr),*), $cancel:ident($($arg3:expr),*)) => {
+        PollI2c {
+            i2c: $me,
+            is_ready: |me| { me.$is_ready($($arg),*) },
+            setup_flags: |me| { me.$setup_flags($($arg2).*); },
+            cancel: |me| {me.$cancel($($arg3),*)},
+            running: true,
+            phantom: PhantomData
+        }
+    };
+}
+
+impl<T, Sda, Scl> I2C<T, (Sda, Scl), Controller<Async>>
+where
+    T: SubsystemReset + Deref<Target = RegisterBlock>,
+    Sda: ValidPinSda<T>,
+    Scl: ValidPinScl<T>,
+{
+    /// Configures the I2C peripheral to work in controller mode and implement async.await APIs.
```

Comment:
Perhaps add a warning regarding cancellation?
```suggestion
    /// Configures the I2C peripheral to work in controller mode and implement async.await APIs.
    ///
    /// Warning: Cancelling a Future returned by this controller (that is, dropping it before `poll` returned `Ready`)
    /// can cause async code to block for relatively long time spans, if the code running on `drop` needs to
    /// terminate running i2c transfers. If your code can't tolerate such delays, make sure to always finish polling
    /// the futures. 
```